package/src/dynamic-env.ts

```diff
@@ -0,0 +1,50 @@
+import { readFileSync } from "node:fs";
+import type { HookParameters } from "astro";
+import { createResolver } from "astro-integration-kit";
+import { hasIntegration } from "astro-integration-kit/utilities";
+import type { Options } from "./integration.js";
+
+type Runtime = Required<Options>["runtime"];
+
+const getRuntime = (
+	config: HookParameters<"astro:config:setup">["config"],
+): Runtime => {
+	const check = (name: string) =>
+		hasIntegration({
+			name,
+			config,
+		});
+
+	if (check("@astrojs/cloudflare")) {
+		return "cloudflare";
+	}
+	if (check("@astrojs/deno")) {
+		return "deno";
+	}
+	if (check("astro-bun-adapter")) {
+		return "bun";
+	}
+	return "node";
+};
+
+export const dynamicEnv = (
+	{ config, addMiddleware }: HookParameters<"astro:config:setup">,
+	providedRuntime?: Runtime,
+) => {
+	const { resolve } = createResolver(import.meta.url);
+
+	const runtime = providedRuntime ?? getRuntime(config);
+
+	addMiddleware({
+		entrypoint: resolve("./middleware.ts"),
+		order: "pre",
+	});
+
+	return {
+		dtsContent: readFileSync(resolve("./stubs/dynamic-env/dts.d.ts"), "utf-8"),
+		virtualImport: readFileSync(
+			resolve(`./stubs/dynamic-env/${runtime}.mjs`),
+			"utf-8",
+		),
```

Comment:
Ah that's interesting! Actually this is a workaround but that will eventually be part of the adapter api. Does it change anything tho? I don't think so. @alexanderniebuhr I'm interested in your thoughts on this

Reply:
In theory I don't think it matters. The only thing which is important, IIUC, is that the variable is set and available in both _(node & other runtime)_ and be named the same. 

The only difference would be how the var is resolved, correct?
The only question is, does it work for runtimes without `fs` if you use `readFileSync` here?

Reply:
That is not the problematic part. The file will be loaded by Node since this run at build time. But if the project tries to read a dynamic variable on a prerendered page it will break.

When deploying to Cloudflare, for example, it will use the code that works on SSR on Cloudflare, but that code doesn't run on Node.

It will break on the route using it, not here

Reply:
Alright so the logic would be
- On build/prerender, use node env resolution
- In SSR, use the adapter logic if defined, otherwise fallback to node env resolution
Is there a way to know what is the current case and use that somehow?

Reply:
> Is there a way to know what is the current case and use that somehow?

There is no easy way for it currently. You can do a runtime check of what runtime are you on between Node and the adapter's runtime by checking `typeof process`. This allows detecting whether it is running on Node in those cases.

Reply:
> checking `typeof process`

That wouldn't work in all cases.. I guess the safer way would be to use Astro's config to see which adapter is set 🤔